src/yki/boundary/cas.clj

```diff
@@ -30,14 +34,40 @@
                                      :headers      {"Caller-Id" caller-id}})))
 
 (defn- json-request [^String method url data]
-  (let [json-str        (json/write-str data)
+  (let [body            (when data (json/write-str data))
         request-builder (RequestBuilder. method)]
     (doto request-builder
       (.addHeader "Content-Type" "application/json")
-      (.setBody json-str)
+      (.setBody body)
       (.setUrl url))
     (.build request-builder)))
 
+(defn- clear-ticket-stores! [^CasClient cas-client]
+  (let [cls                   (.getClass cas-client)
+        session-fetcher-field (.getDeclaredField cls "casSessionFetcher")
+        _                     (.setAccessible session-fetcher-field true)
+        session-fetcher       (.get session-fetcher-field cas-client)]
+    (.clearTgtStore ^CasSessionFetcher session-fetcher)
+    (.clearSessionStore ^CasSessionFetcher session-fetcher)))
+
+(defn- refresh-tickets-response? [url-helper {:keys [headers status]}]
+  (and (= 302 status)
+       (when-let [location (.get ^HttpHeaders headers "Location")]
+         (str/starts-with? location (url-helper :cas.login.root)))))
+
+(defn- cas-http [^CasClient cas-client url-helper method url body]
+  (let [request  (json-request method url body)
+        execute! #(->> request
+                       (.executeBlocking cas-client)
+                       (process-response))
+        response (execute!)]
+    (if (refresh-tickets-response? url-helper response)
+      (do
+        (log/info "Got redirected to CAS-login; refreshing tickets.")
+        (clear-ticket-stores! cas-client)
+        (execute!))
+      response)))
```

Comment:
PR:n pihvi: jos CAS-autentikoituun HTTP-pyyntöön vastataan 302 Redirectillä, kyseessä on oletettavasti kehotus autentikoitua uudestaan kohdepalveluun. Mikäli näin on, pakotetaan uuden ticket granting ticketin sekä session ticketin haku, ja yritetään uudestaan alkuperäisellä pyynnöllä.